alibi_detect/od/pytorch/ensemble.py

```diff
@@ -0,0 +1,341 @@
+
+        Returns
+        -------
+        `Torch.Tensor` of mean of top `k` scores.
+        """
+        if self.k is None:
+            self.k = int(np.ceil(scores.shape[1]/2))
+        sorted_scores, _ = torch.sort(scores, 1)
+        return sorted_scores[:, -self.k:].mean(-1)
+
+
+class AverageAggregator(BaseTransformTorch):
+    def __init__(self, weights: Optional[torch.Tensor] = None):
+        """Averages the scores of the detectors in an ensemble.
+
+        Parameters
+        ----------
+        weights
+            Optional parameter to weight the scores. If `weights` is left ``None`` then will be set to
+            a vector of ones.
+
+        Raises
+        ------
+        ValueError
+            If `weights` does not sum to ``1``.
+        """
+        super().__init__()
+        if weights is not None and weights.sum() != 1:
+            raise ValueError("Weights must sum to 1.")
+        self.weights = weights
+
+    def _transform(self, scores: torch.Tensor) -> torch.Tensor:
+        """Averages the scores of the detectors in an ensemble. If weights were passed in the `__init__`
+        then these are used to weight the scores.
+        ----------
+        scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+
+        Returns
+        -------
+        `Torch.Tensor` of mean of scores.
+        """
+        if self.weights is None:
+            m = scores.shape[-1]
+            self.weights = torch.ones(m, device=scores.device)/m
+        return scores @ self.weights
+
+
+class MaxAggregator(BaseTransformTorch):
+    def __init__(self):
+        """Takes the maximum of the scores of the detectors in an ensemble."""
+        super().__init__()
+
+    def _transform(self, scores: torch.Tensor) -> torch.Tensor:
+        """Takes the maximum score of a set of detectors in an ensemble.
+
+        Parameters
+        ----------
+        scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+
+        Returns
+        -------
+        `Torch.Tensor` of maximum scores.
+        """
+        vals, _ = torch.max(scores, dim=-1)
+        return vals
+
+
+class MinAggregator(BaseTransformTorch):
+    def __init__(self):
+        """Takes the minimum score of a set of detectors in an ensemble."""
+        super().__init__()
+
+    def _transform(self, scores: torch.Tensor) -> torch.Tensor:
+        """Takes the minimum score of a set of detectors in an ensemble.
+
+        Parameters
+        ----------
+        scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+
+        Returns
+        -------
+        `Torch.Tensor` of minimum scores.
+        """
+        vals, _ = torch.min(scores, dim=-1)
+        return vals
+
+
+class Ensembler(BaseFittedTransformTorch):
+    def __init__(self,
+                 normalizer: Optional[BaseFittedTransformTorch] = None,
+                 aggregator: BaseTransformTorch = AverageAggregator()):
```

Comment:
Is this valid or is it going to define the same `aggregator` object for multiple instances of `Ensembler`?